R/chop.R

```diff
@@ -96,27 +96,167 @@ unchop <- function(data, cols, keep_empty = FALSE, ptype = NULL) {
     }
   }
 
-  # If multiple columns, create one data frame for each row
-  # https://github.com/tidyverse/tibble/issues/580
-  x <- pmap(as.list(data)[cols], vec_recycle_common)
-  x <- map(x, ~ new_data_frame(drop_null(.x)))
+  # In case `x` is a grouped data frame and any `cols` are lists,
+  # in which case `[.grouped_df` will error
+  cols <- new_data_frame(unclass(data)[cols])
 
-  n <- map_int(x, vec_size)
-  out <- vec_slice(data, rep(vec_seq_along(data), n))
+  res <- df_unchop_info(cols, ptype)
+  new_cols <- res$val
+  slice_loc <- res$loc
 
-  if (nrow(data) == 0) {
-    new_cols <- map(data[cols], ~ attr(.x, "ptype") %||% unspecified(0))
-  } else {
-    new_cols <- vec_rbind(!!!x, .ptype = ptype)
-  }
+  out <- vec_slice(data, slice_loc)
 
   out <- update_cols(out, new_cols)
   reconstruct_tibble(data, out)
 }
 
-
 # Helpers -----------------------------------------------------------------
 
+# `df_unchop_info()` takes a data frame and unchops every column separately.
+# This preserves the width, but changes the size. Rows that are made entirely of
+# list column elements of `NULL` are dropped. If `x` has any data frame columns,
+# these will be improperly treated as lists until `vec_slice2()` is implemented,
+# but this should be extremely rare.
+
+# `df_unchop_info()` returns a data frame of two columns:
+# - `loc` locations that map each row to their original row in `x`. Generally
+#   used to slice the data frame `x` was subset from to align it with `val`.
+# - `val` the unchopped data frame.
+
+df_unchop_info <- function(x, ptype) {
+  width <- length(x)
+  size <- vec_size(x)
+
+  seq_len_width <- seq_len(width)
+  seq_len_size <- seq_len(size)
+
+  sizes <- rep_len(NA_integer_, size)
+
+  # Gather the common size of each row.
+  # `NULL` elements are ignored in the size calculation by treating their
+  # size as `NA` and then deferring to the size of any other element in the row.
+  # If only `NULL` values are in the row, the `NA` size is finalised to `0`.
+  for (i in seq_len_width) {
+    col <- x[[i]]
+
+    for (j in seq_len_size) {
+      # TODO: col[[j]] -> vec_slice2(col, j)
+      elt <- col[[j]]
+
+      old_size <- sizes[[j]]
+      new_size <- tidyr_size(elt)
+
+      sizes[[j]] <- tidyr_size2(old_size, new_size)
+    }
+  }
+
+  sizes <- map_int(sizes, tidyr_size_finalise)
```

Comment:
It is possible to simplify this and remove `tidyr_size()`, `tidyr_size2()` and `tidyr_size_finalise()` by using

```r
  row <- vector("list", width)

  for (i in seq_len_size) {
    for (j in seq_len_width) {
      row[[j]] <- x[[j]][[i]]
    }

    sizes[[i]] <- vec_size_common(!!!row)
  }
```

but this ends up being a fair bit slower. Maybe if we reimplement in C this will be a faster approach.

Reply:
This is effectively creating a `vec_size(x)` x `length(x)` matrix containing the sizes of each element, and then computing the common size for each row? (Just want to make sure I understand the algorithm here)

Reply:
Yes, that is exactly right and is a nice way to explain it. I've added another code comment to clarify.